### PR TITLE
Refactor GetSpatialIndexOffset

### PR DESCRIPTION
**Pull request by ZehMatt**
Further simplified

### CODE REVIEW

**Review by Gymnasiast — COMMENTED** [review]
Code looks good and also gets rid of a refactor error that sneaked into the previous PR.

Perhaps a good idea if someone tests it.

File: src/openrct2/world/Sprite.cpp [hunk 1]
```diff
@@ -46,21 +46,13 @@ static std::array<std::vector<uint16_t>, SPATIAL_INDEX_SIZE> gSpriteSpatialIndex
...
-    if (x == LOCATION_NULL)
-        return SPATIAL_INDEX_LOCATION_NULL;
-
-    const auto tileX = std::clamp<size_t>(x / COORDS_XY_STEP, 0, MAXIMUM_MAP_SIZE_TECHNICAL);
-    const auto tileY = std::clamp<size_t>(y / COORDS_XY_STEP, 0, MAXIMUM_MAP_SIZE_TECHNICAL);
-    const auto index = tileX * MAXIMUM_MAP_SIZE_TECHNICAL + tileY;
-
-    if (index >= std::size(gSpriteSpatialIndex))
-    {
+    if (loc.IsNull() || !map_is_location_valid(loc))
```
Comment (duncanspumpkin) [inline src/openrct2/world/Sprite.cpp:51]: Do we really need the additional `map_is_location_valid`? This function is going to be called super super often.

Comment (ZehMatt) [reply src/openrct2/world/Sprite.cpp:51]: It no longer clamps so it needs to check that.

Comment (ZehMatt) [reply src/openrct2/world/Sprite.cpp:51]: It turns out that this new form is also the fastest, see https://quick-bench.com/q/fHtaplRpsvlpiWH3dvulMwgR90A

**Review by Broxzier — APPROVED** [review]

**Comment by Broxzier** [comment]
Weren't negative coordinates already filtered out by calling `map_is_location_valid`? Why is this call not necessary anymore?

**Comment by ZehMatt** [comment]
> Weren't negative coordinates already filtered out by calling `map_is_location_valid`? Why is this call not necessary anymore?

They should not be filtered, they are valid coordinates. Based on the rotation it may turn a component negative its not really out of bounds.

**Review by Gymnasiast — APPROVED** [review]
Testing revealed no bugs.

Would like @duncanspumpkin to confirm this also solves the performance issue.

**Comment by duncanspumpkin** [comment]
I've locally applied this ontop of another branch and it resolved my perfomance issues. I think though we should also add a 
```cpp
    if (!map_is_location_valid(pos))
    {
        return;
    }
```
to the sprite_paint_setup function to avoid further pointless work.

**Comment by ZehMatt** [comment]
> I've locally applied this ontop of another branch and it resolved my perfomance issues. I think though we should also add a
> 
> ```c++
>     if (!map_is_location_valid(pos))
>     {
>         return;
>     }
> ```
> 
> to the sprite_paint_setup function to avoid further pointless work.

I'm pretty sure this will break the rendering.

**Comment by duncanspumpkin** [comment]
There should be no entities outside of `map_is_location_valid` that require rendering. Lets discuss it outside of this as this fixes one bit either way.

**Pull request merged**